fs/remote/resolver.go

```diff
@@ -271,8 +271,15 @@ func (tr *transport) RoundTrip(req *http.Request) (*http.Response, error) {
 	if resp.StatusCode == http.StatusUnauthorized {
 		log.G(ctx).Infof("Received status code: %v. Refreshing creds...", resp.Status)
 
-		// prepare authorization for the target host using docker.Authorizer
-		if err := tr.auth.AddResponses(ctx, []*http.Response{resp}); err != nil {
+		// prepare authorization for the target host using docker.Authorizer.
+		// The docker authorizer only refreshes OAuth tokens after two
+		// successive 401 errors for the same URL. Rather than issue the same
+		// request multiple times to tickle the token-refreshing logic, just
+		// provide the same response twice to trick it into refreshing the
+		// cached OAuth token.
+		// TODO: fix after https://github.com/containerd/containerd/pull/8735
+		// is merged and released.
+		if err := tr.auth.AddResponses(ctx, []*http.Response{resp, resp}); err != nil {
```

Comment:
Can you confirm this fixes the issue? This AddResponses call will internally delete the "auth handler" but not setup a new one which means the next `Authorize` call will not have a token associated with it. I think for this hack to work you'd need to call `AddResponses` twice, once to delete the current auth handler and a second time to create a new one.

Reply:
It does work due to higher-level retry logic, but I think it's wiser to explicitly invalidate-and-refresh it here like you describe, so I changed it to do so.

Reply:
SGTM.